File: packages/ilc-py/src/devalbo_ilc/hosts/test_host.py
from __future__ import annotations

from dataclasses import dataclass, field

from devalbo_ilc.generated.types import ConsoleIo, Environment
from devalbo_ilc.hosts.callbacks import ConsoleIoCallbacks, console_io_from_callbacks


@dataclass
class LogEntry:
    level: str
    message: str


@dataclass
class TestEnvironment:
    stdin: list[str] = field(default_factory=list)
    logs: list[LogEntry] = field(default_factory=list)
    _stdin_index: int = 0
    _console_io: ConsoleIo | None = None

    async def consoleIo(self) -> ConsoleIo:
        if self._console_io is None:
            self._console_io = console_io_from_callbacks(
                ConsoleIoCallbacks(
                    on_info=lambda message: self._log("info", message),
                    on_error=lambda message: self._log("error", message),
                    on_read_line=self._read_line,
                )
            )
        return self._console_io

    def _log(self, level: str, message: str) -> None:
        self.logs.append(LogEntry(level=level, message=message))

    def _read_line(self) -> str | None:
        if self._stdin_index >= len(self.stdin):
            return None
        line = self.stdin[self._stdin_index]
        self._stdin_index += 1
        return line if line else None


def create_test_environment(*, stdin: list[str] | None = None) -> TestEnvironment:
    return TestEnvironment(stdin=list(stdin or []))

File: packages/ilc-rs/src/hosts/test_host.rs
use std::sync::{Arc, Mutex};

use async_trait::async_trait;

use crate::generated::types::{ConsoleIo, Environment};
use crate::hosts::shared::box_console_io;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LogEntry {
    pub level: &'static str,
    pub message: String,
}

struct TestConsoleIo {
    logs: Arc<Mutex<Vec<LogEntry>>>,
    stdin: Arc<Mutex<Vec<String>>>,
    stdin_index: Arc<Mutex<usize>>,
}

#[async_trait]
impl ConsoleIo for TestConsoleIo {
    async fn info(&self, message: &str) {
        self.logs.lock().unwrap().push(LogEntry {
            level: "info",
            message: message.to_string(),
        });
    }

    async fn error(&self, message: &str) {
        self.logs.lock().unwrap().push(LogEntry {
            level: "error",
            message: message.to_string(),
        });
    }

    async fn readLine(&self) -> Option<String> {
        let mut index = self.stdin_index.lock().unwrap();
        let stdin = self.stdin.lock().unwrap();
        if *index >= stdin.len() {
            return None;
        }
        let line = stdin[*index].clone();
        *index += 1;
        if line.is_empty() { None } else { Some(line) }
    }
}

pub struct TestEnvironment {
    console: Arc<dyn ConsoleIo + Send + Sync>,
    pub logs: Arc<Mutex<Vec<LogEntry>>>,
}

#[async_trait]
impl Environment for TestEnvironment {
    async fn consoleIo(&self) -> Box<dyn ConsoleIo + Send + '_> {
        box_console_io(self.console.clone())
    }
}

pub fn create_test_environment(stdin: Vec<String>) -> TestEnvironment {
    let logs = Arc::new(Mutex::new(Vec::new()));
    let console = Arc::new(TestConsoleIo {
        logs: logs.clone(),
        stdin: Arc::new(Mutex::new(stdin)),
        stdin_index: Arc::new(Mutex::new(0)),
    });
    TestEnvironment { console, logs }
}
